syn/flat_encode.go

```diff
@@ -184,7 +184,7 @@ func (e *encoder) encodeTermTag(tag byte) error {
 }
 
 func (e *encoder) safeEncodeBits(numBits byte, val byte) error {
-	if 2<<numBits <= val {
+	if val >= (1 << numBits) {
```

Comment:
Is this actually correct? Does changing the order/operator of the comparison here require changing the number of bits that were shifting?

Reply:
So... there's a lot of math explanation here, but the answer is

> Answer to Your Question
> No, the number of bits doesn't need to change. The original condition was using the wrong mathematical threshold. The bit count numBits is correct - the issue was that `2<<numBits` was checking against `2^(numBits+1)` instead of `2^numBits`.
>
> This was a critical bug fix that prevented silent truncation of values that don't > fit in the allocated bit space, which could lead to incorrect Plutus term encoding.